### PR TITLE
refactor: Geometry hierarchy material map

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This PR modify the JsonGeometryConverter that is used in the material mapping process to use the geometry hierarchy this give us a more uniform Json output compare to other Acts json files.

The python script used to configure which surface with which binning to use for the material mapping have been modify to accommodate the file structure.

I modify the key used for the ID in GeometryHierarchyMapJsonConverter.hpp, with the previous key the element would be order alphabetically in the file (which made them extremely hard to read). In the same way some key used in the writing of the material mapping use Capital letter, this is not a mistake it is to for the element to be ordered in a more readable way in the files.

Added 2 files with to_json and from_json, the first one is used to convert material, material surface and material volume. The second convert pointer on surface and tracking volume to initialise the material map json file used to configure the mapping surface/volume and binning. This shouldn't conflict with #662 but we should keep an eye on it.

### CODE REVIEW

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp [hunk 1]
```diff
@@ -74,30 +74,30 @@ class GeometryHierarchyMapJsonConverter {
     nlohmann::json encoded;
     // only store non-zero identifiers
     if (id.volume()) {
-      encoded["volume"] = id.volume();
```
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:77]: Why did you change the way the `GeometryHierarchyMapJsonConverter` is writing out the geometry tags?

File: Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <string>
+
+namespace Acts {
+
+/// @struct jsonKey
+///
+/// @brief store in a single place the different key used for the material
+/// mapping
+struct jsonKey {
```
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp:19]: O, I think we need some naming convention here, I was writing everything in small (no camel case either), because of being an input file, the keys should follow CamelCase though.
```c++
std::string nameKey = "name";
```
That would be my suggestion?

File: Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp [hunk 3]
```diff
@@ -1,19 +1,17 @@
 // This file is part of the Acts project.
 //
```
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp:2]: Also here, we are mixing

`ThingJsonConverter.hpp` and `JsonThingConverter.hpp` - I am good either way, but after the `SurfaceJsonConverter.hpp` is in, you should either synchronize either.

File: Plugins/Json/src/SurfaceJsonConverter.cpp [hunk 4]
```diff
@@ -25,6 +27,15 @@
 #include "Acts/Surfaces/StrawSurface.hpp"
 #include "Acts/Surfaces/TrapezoidBounds.hpp"
 
+void Acts::to_json(nlohmann::json& j, const Acts::SurfaceAndMaterial& surface) {
+  to_json(j, surface.first);
```
Comment (asalzburger) [inline Plugins/Json/src/SurfaceJsonConverter.cpp:31]: shouldn't this be `(*surface.first)` ?
I only impelemted the surface converter for reference and `shared_ptr`

File: Plugins/Json/src/SurfaceJsonConverter.cpp [hunk 5]
```diff
@@ -25,6 +27,15 @@
...
+  to_json(j, surface.second.get());
+}
+
+void Acts::to_json(nlohmann::json& j, const Acts::SurfacePointer& surface) {
+  to_json(j, *surface);
```
Comment (asalzburger) [inline Plugins/Json/src/SurfaceJsonConverter.cpp:36]: Oh, now I understand ... I think if the `SurfaceJsonConverter` from the framework is in, this gets easier. I added another convenience function to write out a `const Surface& ref`

File: Tests/Data/geometry-hierarchy-map.json [hunk 6]
```diff
@@ -12,24 +12,24 @@
             }
         },
         {
-            "volume": 1,
-            "layer": 2,
+            "Volume": 1,
```
Comment (asalzburger) [inline Tests/Data/geometry-hierarchy-map.json:15]: I liked the original "volume" , "layer" better

**Review by asalzburger — CHANGES_REQUESTED** [review]
Can you please update this once #679 is in?
It adds a convenience function for the `SurfaceJsonConverter`

File: Tests/Data/geometry-hierarchy-map.json (hunk 6 above)
Comment (Corentin-Allaire) [reply Tests/Data/geometry-hierarchy-map.json:15]: Me too but the issue is that when you write the json file the key are ordered alphabetically which mean it use to look like : 
```
            "layer": 4,
            "sensitive": 5,
            "value": {
                "x": 3.25,
                "y": -5
            }
            "volume": 3,
``` 
Which was confusing. I don't know if there is a way to force the writing to not order the key.

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 1 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:77]: As I explain in the description of the PR the key are ordered alphabetically hen writing the file. I used underscore and uppercase  to force the ordering of the key to be :
volume
boundary
layer
approach
sensitive

File: Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp (hunk 2 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp:19]: Here is the same idea as before I used upper case for the information I wanted to appear at the top when the json object is written. For me when you look at the file the first thing you want to see is the name and ids of the objct that is why I used uppercases

File: Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp (hunk 3 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp:2]: I agree I should change this name. Let me thing of something better

File: Tests/Data/geometry-hierarchy-map.json (hunk 6 above)
Comment (Corentin-Allaire) [reply Tests/Data/geometry-hierarchy-map.json:15]: this now fixed with PR #686

File: Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp (hunk 1 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryHierarchyMapJsonConverter.hpp:77]: this now fixed with PR #686

File: Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp (hunk 2 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp:19]: What is your opinion on this one @asalzburger ? We cannot use the ordered_json here since it would get converted to normal json by the to_json function. I don't mind having everything be in lower case but I also find it nice to have some useful information at the top of the object in the json file (but that require an uppercase).

File: Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp (hunk 3 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/JsonGeometryConverter.hpp]: I changed it to MapJsonConverter

File: Plugins/Json/include/Acts/Plugins/Json/MapJsonConverter.hpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+using SurfaceAndMaterial =
+    std::pair<std::shared_ptr<const Acts::Surface>,
+              std::shared_ptr<const Acts::ISurfaceMaterial>>;
+using TrackingVolumeAndMaterial =
+    std::pair<const Acts::TrackingVolume*,
+              std::shared_ptr<const Acts::IVolumeMaterial>>;
+
+/// @class MapJsonConverter
+///
+/// @brief read the material from Json
```
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/MapJsonConverter.hpp:34]: This only converts MaterialMaps, why not 'MaterialMapJsonConverter' ?

File: Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp (hunk 2 above)
Comment (asalzburger) [reply Plugins/Json/include/Acts/Plugins/Json/GeometryJsonKeys.hpp:19]: Ok, we could have it all Upper case, such as "NAME".

File: Plugins/Json/include/Acts/Plugins/Json/MaterialJsonConverter.hpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
```
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/MaterialJsonConverter.hpp:8]: Perfect, I like this.

File: Tests/UnitTests/Plugins/Json/MaterialMapJsonConverterTests.cpp [hunk 9]
```diff
@@ -8,7 +8,7 @@
```
Comment (asalzburger) [inline Tests/UnitTests/Plugins/Json/MaterialMapJsonConverterTests.cpp:8]: The test is even called `MaterialMapJsonConverterTests` - so the class and file should be called `MaterialMapJsonConverter`.

**Review by asalzburger — CHANGES_REQUESTED** [review]
Minor changes: rename to `MaterialMapJsonConverter` and yes, let's change the name key to 'NAME' to have it on top.

File: Plugins/Json/include/Acts/Plugins/Json/MapJsonConverter.hpp (hunk 7 above)
Comment (Corentin-Allaire) [reply Plugins/Json/include/Acts/Plugins/Json/MapJsonConverter.hpp:34]: Yeah you are right I will change that

**Review by asalzburger — APPROVED** [review]
